From: Rob Evers <revers@redhat.com>
Date: Wed, 27 Oct 2010 15:23:15 -0400
Subject: [scsi] lpfc: fail io w/lost frame and target check cond
Message-id: <1288192999-24221-24-git-send-email-revers@redhat.com>
Patchwork-id: 28948
O-Subject: [RHEL5.6 PATCH 23/27] lpfc: Fail I/O when adapter detects a lost
	frame and target reports a check condition
Bugzilla: 639028

lpfc: Fail I/O when adapter detects a lost frame and target reports a check condition

From: Rob Evers on behalf of Emulex <revers@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=636289

diff --git a/drivers/scsi/lpfc/lpfc_scsi.c b/drivers/scsi/lpfc/lpfc_scsi.c
index a17f634..dc3448e 100644
--- a/drivers/scsi/lpfc/lpfc_scsi.c
+++ b/drivers/scsi/lpfc/lpfc_scsi.c
@@ -1493,17 +1493,24 @@ lpfc_handle_fcp_err(struct lpfc_hba *phba, struct lpfc_vport *vport,
 	 * Check SLI validation that all the transfer was actually done
 	 * (fcpi_parm should be zero). Apply check only to reads.
 	 */
-	} else if ((scsi_status == SAM_STAT_GOOD) && fcpi_parm &&
-			(cmnd->sc_data_direction == DMA_FROM_DEVICE)) {
-		lpfc_printf_log(phba, KERN_WARNING,
-				LOG_FCP | LOG_FCP_ERROR,
-				"(%d):0734 FCP Read Check Error Data: "
-				"x%x x%x x%x x%x\n",
-				(vport ? vport->vpi : 0),
-				be32_to_cpu(fcpcmd->fcpDl),
-				be32_to_cpu(fcprsp->rspResId),
-				fcpi_parm, cmnd->cmnd[0]);
-		host_status = DID_ERROR;
+	} else if (fcpi_parm && (cmnd->sc_data_direction == DMA_FROM_DEVICE)) {
+		lpfc_printf_vlog(vport, KERN_WARNING, LOG_FCP | LOG_FCP_ERROR,
+				 "9029 FCP Read Check Error Data: "
+				 "x%x x%x x%x x%x x%x\n",
+				 be32_to_cpu(fcpcmd->fcpDl),
+				 be32_to_cpu(fcprsp->rspResId),
+				 fcpi_parm, cmnd->cmnd[0], scsi_status);
+		switch (scsi_status) {
+		case SAM_STAT_GOOD:
+		case SAM_STAT_CHECK_CONDITION:
+			/* Fabric dropped a data frame. Fail any successful 
+			 * command in which we detected dropped frames. 
+			 * A status of good or some check conditions could
+			 * be considered a successful command.
+			 */
+			host_status = DID_ERROR;
+			break;
+		}
 		cmnd->resid = cmnd->request_bufflen;
 	}
 
